src/main/java/net/fabricmc/loader/discovery/ModResolver.java

```diff
@@ -340,20 +340,56 @@ private void addErrorToList(ModCandidate candidate, ModDependency dependency, Ma
 			errors.append("\n - Mod ").append(candidate.getInfo().getId()).append(" ").append(errorType).append(" mod ").append(dependency).append(", ");
 
 			if (depCandidate == null) {
-				errors.append("which is missing");
-			} else if (cond) {
-				errors.append("but a different version is present: ").append(depCandidate.getInfo().getVersion());
-			} else if (errorType.contains("conf")) {
-				// CONFLICTS WITH
-				errors.append("but the conflicting version is present: ").append(depCandidate.getInfo().getVersion());
+				errors.append("which is missing!");
+				errors.append("\n     You must install ").append(horribleModDependencyToStringMethod(dependency)).append(" of ").append(depModId).append(".");
 			} else {
-				errors.append("but the breaking version is present: ").append(depCandidate.getInfo().getVersion());
+				final String depCandidateVer = depCandidate.getInfo().getVersion().getFriendlyString();
+				if (cond) {
+					errors.append("but a different version is present: ").append(depCandidateVer).append("!");
+					errors.append("\n     You must install ").append(horribleModDependencyToStringMethod(dependency)).append(" of ").append(depModId).append(".");
+				} else if (errorType.contains("conf")) {
+					// CONFLICTS WITH
+					errors.append("but the conflicting version is present: ").append(depCandidateVer).append("!");
+					errors.append("\n     While this won't prevent you from starting the game, the developer(s) of ").append(candidate.getInfo().getId());
+					errors.append("\n     have found that version ").append(depCandidateVer).append(" of ").append(depModId);
+					errors.append("\n     conflicts with their mod. It is heavily recommended to remove one of the mods.");
+				} else {
+					errors.append("but the breaking version is present: ").append(depCandidate.getInfo().getVersion()).append("!");
+					errors.append("\n     The developer(s) of ").append(candidate.getInfo().getId());
+					errors.append("\n     have found that version ").append(depCandidateVer).append(" of ").append(depModId);
+					errors.append("\n     critically conflicts with their mod. You must remove one of the mods.");
+				}
 			}
 
-			errors.append("!");
+			//errors.append("!");
 		}
 	}
 
+	// lord forgive me for what I must do
+	private static String horribleModDependencyToStringMethod(ModDependency dependency) {
+		String depStr = dependency.toString();
+		int verStart = depStr.indexOf('[');
+		if (verStart < 0)
+			return "unknown version";
+		int verEnd = depStr.indexOf(']');
+		if (verEnd < verStart)
+			return "unknown version";
+		String verStr = depStr.substring(verStart + 1, verEnd);
+		if ("*".equals(verStr))
```

Comment:
good lord switch block may be least messy

Reply:
should probably add a method on `ModDependency` to do this, really, since `toString` is implementation-independent
case in point: the current mess always returns "unknown version" for v0 mods

Reply:
I did the thing, also it's a switch block now
for v0 mods: https://github.com/FabricMC/fabric-loader/pull/277/files#diff-3239f64a8e4a83838362f26163c22eedR333 (REQUIRES MORE WORK)
for v1 mods: https://github.com/FabricMC/fabric-loader/pull/277/files#diff-e7a80560353754b29221f3af67d286fcR303